In [4]:
import numpy as np
import scipy

**Z-критерий для несвязанных выборок**

Для проверки гипотезы о равентсве долей используется критерий аналогично критерию Стьюдента t.

**z** = $ \Large \frac{\text{Разность выборочных долей}} {\text{Стандартная ошибка разности выборочных долей}}  = \frac{\hat{p_1}-\hat{p_2}} {{S_{\hat{p_1} - \hat{p_2}}}}  = \frac{\hat{p_1}-\hat{p_2}} { \sqrt{ S_{\hat{p_1}}^2 +  S_{\hat{p_2}}^2 } } = \frac{\hat{p_1}-\hat{p_2}} { \sqrt{ \frac{\hat{p_1}(1-\hat{p_1})} {n_1} +  \frac{\hat{p_2}(1-\hat{p_2})} {n_2} } } $


**z**  $ \Large = \frac{|\hat{p_1}-\hat{p_2}| - \frac{1} {2} (\frac{1} {n_1} + \frac{1} {n_2} ) } { \sqrt{ \frac{\hat{p_1}(1-\hat{p_1})} {n_1} +  \frac{\hat{p_2}(1-\hat{p_2})} {n_2} } }$

Корректно использовать в том случае, когда n*p и n*(1 − p) превосходят 5 для всех групп.

Пример:
**Задача.** DC работает над новым баннером, который хочет разместить на одной из своих рекламных площадок. Было придумано 2 варианта баннеров (A и B). На первый баннер за первые сутки кликнуло 100 человек из 300 посетителей, на второй баннер 400 человек из 800 посетителей. Решить,какой баннер оставлять.

Гипотезы:

H0: доля кликов по двум баннерам одинакова

H1: доля кликов по баннеру А меньше, чем по баннеру B

In [2]:
from statsmodels.stats.proportion import proportions_ztest
count = np.array([100, 300])
nobs = np.array([300, 800])
proportions_ztest(count, nobs, alternative='smaller')

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


(-1.279415789297898, 0.10037533831983553)

Здесь второе значение 0.10037.. это p-value или уровень значимости.
Уровень значимости — допустимая для данной задачи вероятность ошибки первого рода (ложноположительного решения, false positive), то есть вероятность отклонить нулевую гипотезу, когда на самом деле она верна.

Обычно считают, что если p-value > 0.5, то нельзя отвергать нулевую гипотезу. Чем больше p-value, тем больше тест уверен, что гипотеза правильная. В итоге, исходная гипотеза подтвердилась.

Текущая задача:
Изменилось ли количество упоминаний Витамишки с 23 февраля?

H0: количество упоминаний одинаково

H1: количество упоминаний после 23 февраля меньше

In [3]:
first = np.array([1716, 2172])
second = np.array([610, 763])
proportions_ztest(second, first, alternative='smaller')

(0.27134939349587817, 0.6069388397888845)

Т. е. тест с 60% вероятностью говорит, что нельзя отвергнуть нулевую гипотезу или другими словами тест сильно уверен, что количество упоминаний не изменилось

**Критерий Хи-квадрат**

In [14]:
from scipy.stats import chi2_contingency
import pandas as pd

In [17]:
data = pd.DataFrame()
data['до 23'] = [1716, 2172 - 1716, 2172]
data['после'] = [610, 763 - 610, 763]
data.index = ['упоминаний Витамишки', 'другое','Итого']
data['Всего'] = data['до 23'] + data['после']
data

до 23  после  Всего
упоминаний Витамишки   1716    610   2326
другое                  456    153    609
Итого                  2172    763   2935

При работе с критерием хи-квадрат встречается понятие, как таблица ожидаемых значений.

Примем нулевую гипотезу H0, что упоминаемость Витамишек одинакова, тогда:

In [18]:
yp_total = 2172/2935
non_yp_total = 763/2935

In [19]:
first_yp = yp_total * 2326
second_yp = yp_total * 609

first_non_yp = non_yp_total * 2326
second_non_yp = non_yp_total * 609

In [20]:
data_E = pd.DataFrame()
data_E['до 23'] = [first_yp,second_yp,2172]
data_E['после'] = [first_non_yp, second_non_yp,763]
data_E.index = ['упоминаний Витамишки', 'другое','Итого']
data_E['Всего'] = data_E['до 23'] + data_E['после']
data_E

до 23      после   Всего
упоминаний Витамишки  1721.31925  604.68075  2326.0
другое                 450.68075  158.31925   609.0
Итого                 2172.00000  763.00000  2935.0

$ \Large \chi^2  = \sum {   \frac{(O - E)^2} {E}  } $, O  - наблюдаемое значение, E - ожидаемое значение.

$ \Large \chi^2  = \sum {   \frac{|O - E| - \frac{1} {2}} {E}  } $ (с поправкой Йеитса - только для таблиц 2х2)

In [22]:
chi2, p, dof, ex = chi2_contingency(data.iloc[:2,:2])

In [23]:
chi2,p,dof,ex

(0.25013354818703054,
 0.6169810577735972,
 1,
 array([[1721.31925043,  604.68074957],
        [ 450.68074957,  158.31925043]]))

In [24]:
p

0.6169810577735972

Критерий Хи-квадрат дает p-value 0.61, т е нулевая гипотеза, что количество упоминаний витамишек не изменилось не может быть отвергнута.